### ISSUE TITLE
Simplified FocusCycler API

### ISSUE THREAD

**Issue by pjasiun** [ckeditor/ckeditor5-ui#259]
First of all, that very cool that we have a tool like this. However, I have some doubts about the API and the fact that `KeystrokeHandler` can be passed there.

It's not the responsibility of the `FocusCycler` to handle keystrokes. It's `KeystrokeHandler` responsibility. Right now it another API to learn. I need to learn that I need to pass both `keystrokeHandler` and actions (it will fail silently otherwise). And this is additional API/code to maintain. For instance: what if I want to add more keystrokes for  `focusNext` action? Can I pass array there?

Instead of:
```js
let focusTracker = new FocusTracker();
let keystrokes = new KeystrokeHandler();
let focusables = new ViewCollection();

let focusCycler = new FocusCycler( {
	focusables: focusables,
	focusTracker: focusTracker,
	keystrokeHandler: keystrokes,
	actions: {
		focusPrevious: 'shift + tab',
		focusNext: 'tab'
	}
} );
```

I would keep it simple remove this API and go with:

```js
let focusTracker = new FocusTracker();
let focusables = new ViewCollection();

let focusCycler = new FocusCycler( { focusables,  focusTracker } );

let keystrokes = new KeystrokeHandler();
let keystrokes.set( 'tab', this._focusCycler.focusNext() );
let keystrokes.set( 'shift + tab', this._focusCycler.focusPrevious() );
```

Also, note that the second case it simple to debug. In the first one, I have no good place to put the `debugger` statement.

**Comment by oskarwrobel** [comment]
I like this idea. IMO `FC` initialization is complex.

**Comment by Reinmar** [comment]
The correct code, for now, would look like this:

```js
const focusTracker = new FocusTracker();
const focusables = new ViewCollection();

const focusCycler = new FocusCycler( { focusables,  focusTracker } );

const keystrokes = new KeystrokeHandler();
const keystrokes.set( 'tab', ( evt, cancel ) => {
	focusCycler.focusNext();
	cancel();
} );
const keystrokes.set( 'shift + tab', ( evt, cancel ) => {
	focusCycler.focusPrevious();
	cancel();
} );
```

Which means that it becomes a bit longer than the declarative actions syntax. But it still makes sense to not mix keystroke handler and focus cycler.

We may also think whether keystroke handler shouldn't cancel the event by default... It seems that in 90% of cases it's what you're expecting.

**Comment by Reinmar** [comment]
And there's also a problem with rewriting such a setting:

```js
				// Navigate toolbar items backwards using the arrow[left,up] keys.
				focusPrevious: [ 'arrowleft', 'arrowup' ],

				// Navigate toolbar items forwards using the arrow[right,down] keys.
				focusNext: [ 'arrowright', 'arrowdown' ]
```

So, maybe `FocusCycler` could expose some function which quickly adds such a keystrokes? Like:

```js
focusCycler.setKeystrokes( keystrokeHandler, {
    focusPrevious: [ 'arrowleft', 'arrowup' ],
    focusNext: [ 'arrowright', 'arrowdown' ]
} );
```

And we're back to a monstrous config, but split into two pieces ;)

**Comment by pjasiun** [comment]
It's for sure better, but don't like such way, because it's not debuggable. I mean, if I code using `keystrokes.set` I can put debugger there, and check if it works or not. I can also check if `focusCycler.focusNext()` do what I want, and know where I suppose to look for the bug. If there is a config I can only look on it as long as I see a mistake. And, it might be just me, but it's sometimes very hard to find it.

I also agree that `keystrokes.set` should cancel by default, since it's what 90% cases need. `keystrokes.set` should be just a util for the most popular cases. For the remaining 10%, keystroke handler could fire an event when a key is pressed we can be handled without preventing. Or they can use `parseKeystroke` util. It's not like, you can't do it another way, so `keystrokes.set` should let you cancel or not.

**Comment by pjasiun** [comment]
In fact, as I think about it, this code: https://github.com/ckeditor/ckeditor5-ui/blob/6f5ec0d0ca25be90de9efac12a92024016a0bba0/src/focuscycler.js#L103-L118 is independent of the focus cycler logic. It could go to any feature which needs to handle keystrokes, so this should handle by the `KeystrokeHandler`, most probably as `keystrokes.set` behavior described above.

**Comment by oskarwrobel** [comment]
> We may also think whether keystroke handler shouldn't cancel the event by default...

Isn't it risky?

**Comment by oskarwrobel** [comment]
I can imagine that someone will waste a lot of time debugging not working event because one of the child views will call `stopPropagation` on `Esc` with no reason.

**Comment by Reinmar** [comment]
> I can imagine that someone will waste a lot of time debugging not working event because one of the child views will call stopPropagation on Esc with no reason.

Yeah. We'll replace non-debuggable actions format with a counterintuitive event cancelling (which only happens by default in this one place) :D. 

This is the tradeoff – more concise API is usually more magical and harder to debug.

But I'm not saying that we won't be able to find some option for keystroke handler to cancel its events automatically when a listener is defined by `set()` and to allow for passing them through (e.g. when you use `on()`).

Still, if you need to set 6 keystrokes (up, left, shift+tab doing "focus previous" and the other keys doing "focus next") using keystroke handler directly is a loooot of repeatable code. We need to be smart here and a simple actions format is the smartest option I can see at the moment.

**Comment by Reinmar** [comment]
Besides, let's not overexaggerate the whole thing – it takes you 2 minutes to learn the actions format. There are far more hardcore things in our API and when we're trying to make things like this **rarely** used API idiot-proof we're just wasting everyone's time.

So, after reconsidering this, the only thing I can agree at the moment is getting rid of `keystrokeHandler` and `actions` params from the main option object and moving them to `setKeystrokes()`. This will decouple the `FC` and `KH` far enough (you'll be able to use `FC` without `KH` at all).
